plugins/uptimerobot/src/hooks/useAutoUpdatingRequest.ts

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2021 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ApiResponse } from '../types';
+import { uptimerobotApiRef } from '../api';
+import { useApi } from '@backstage/core';
+import { useAsync, useInterval } from 'react-use';
+import { useState } from 'react';
+
+export function useAutoUpdatingRequest(apiMethod: Function) {
+  const uptimerobotApi = useApi(uptimerobotApiRef);
+  const [rand, setRand] = useState(0);
+
+  const { value, loading, error } = useAsync(async (): Promise<ApiResponse> => {
```

Comment:
you probably want to use `useAsyncFn` instead, and a `useInterval` to trigger its returned fn over and over again. those are both in react-use

Reply:
I would then additionally need something to trigger the apiMethod initially on page load. So it could look something like this:
```ts
const [ { value, loading, error }, doApiMethod ] = useAsyncFn(async () => {
  return await apiMethod();
}, [ apiMethod ])

useInterval(() => {
  doApiMethod()
}, uptimerobotApi.getUpdateInterval() * 1000)

useEffect(() => {
  doApiMethod()
}, []);
```

However, having an empty dependency array for useEffect gets marked by ESLint. ESLint requires doApiMethod as depdency which leads to an endless loop.

Reply:
Side note: I've been out sick for a bit. We'll respond to your comments soon :)

Reply:
It seems like this has to do with they way I'm passing a function (the `apiMethod`)  to this hook.
I'm calling the hook like this:
```tsx
const { value, loading, error } = useAutoUpdatingRequest(() =>
  uptimerobotApi.getSingleMonitor(entity),
);
```
I'm doing this so that I can pass a property (`entity`) to that `getSingleMonitor`. Also this creates an anonymous function so that I can call `this.fetch` within the class `UptimerobotApi` of that method (see here: https://github.com/backstage/backstage/pull/5524/files#diff-4380c33409d14fb4ff5dcf0618bf03a9c8a17b8061c81275b11b4e8b77fd30deR57).
Apparently because of the re-rendering of that the component containing this hook, this will re-create that anonymous function every time and in combination with the stuff inside the hook therefore results in that loop for that `useEffect`.

I'm not sure what to do here except disabling the linter for that empty dependency array or keeping it like it currently is 🙃

Reply:
Yep you can wrap the function in useCallback (at the point of creation) with the right dependencies, that way it gets memoized

Reply:
Hooks are awesome in a lot of ways - but they sure can be fiddly...

Reply:
Oh, I didn't know that. That works. Thank you :)